### PR TITLE
СимволыОступа в ЗаписьXML.

### PR DESCRIPTION
**Pull request by dmpas**

### CODE REVIEW

**Comment by nixel2007** [comment]
SonarQube analysis reported 1 issue

1. ![MAJOR][MAJOR] [XmlWriterImpl.cs#L27](https://github.com/EvilBeaver/OneScript/blob/9433d85a90ba786aef5a1f78b04d9b4363a6d9fd/src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs#L27): Remove the unused private field 'INDENT_SIZE'. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS1144)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs [hunk 1]
```diff
@@ -247,33 +229,52 @@ public IValue Close()
...
-            if (addBOM == null)
-                enc = TextEncodingEnum.GetEncodingByName(encoding, true);
+        private void ApplySettings(IValue encoding)
+        {
+            var rawEncoding = encoding?.GetRawValue();
+            if (rawEncoding is XmlWriterSettingsImpl)
+            {
+                _settings = rawEncoding as XmlWriterSettingsImpl;
+            }
+            else if ((encoding?.DataType ?? DataType.String) == DataType.String)
```
Comment (EvilBeaver) [inline src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs]: Как-то не улавливаю... в приватном методе параметр encoding который не то settings, не то indent...

Comment (dmpas) [reply src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs]: @EvilBeaver если посмотришь СП, то там в ОткрытьФайл или УстановитьСтроку передаётся либо кодировка, либо ПараметрыЗаписиXML. Соответственно, приватная метода просто отрабатывает этот параметр.

Comment (EvilBeaver) [reply src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs]: Может тогда параметр переименовать в encodingOrSettings?

Comment (dmpas) [reply src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs]: да, пожалуй.

**Comment by dmpas** [comment]
https://github.com/EvilBeaver/OneScript/issues/768#issuecomment-397848410
~~1. Не пишется BOM~~
2. Пробел перед закрывающим слэшем

**Comment by nixel2007** [comment]
SonarQube analysis reported 1 issue

1. ![INFO][INFO] [XmlWriterImpl.cs#L321](https://github.com/EvilBeaver/OneScript/blob/30d2725fe29dd66ce115b5e7d51649f0ebc30c95/src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs#L321): Complete the task associated to this 'TODO' comment. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by dmpas** [comment]
~~2. Пробел перед закрывающим слэшем~~
3. Лишний перенос строки при ЗаписатьСтроку

**Comment by nixel2007** [comment]
SonarQube analysis reported 6 issues

1. ![CRITICAL][CRITICAL] [ContextValuesMarshaller.cs#L53](https://github.com/EvilBeaver/OneScript/blob/30d2725fe29dd66ce115b5e7d51649f0ebc30c95/src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs#L53): Refactor this method to reduce its Cognitive Complexity from 20 to the 15 allowed. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS3776)
1. ![MAJOR][MAJOR] [ContextValuesMarshaller.cs#L217](https://github.com/EvilBeaver/OneScript/blob/30d2725fe29dd66ce115b5e7d51649f0ebc30c95/src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs#L217): Remove this useless assignment to local variable 'result'. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS1854)
1. ![MAJOR][MAJOR] [XmlWriterImpl.cs#L19](https://github.com/EvilBeaver/OneScript/blob/30d2725fe29dd66ce115b5e7d51649f0ebc30c95/src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs#L19): Fix this implementation of 'IDisposable' to conform to the dispose pattern. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS3881)
1. ![MAJOR][MAJOR] [XmlWriterImpl.cs#L24](https://github.com/EvilBeaver/OneScript/blob/30d2725fe29dd66ce115b5e7d51649f0ebc30c95/src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs#L24): Make '_nsmap' 'readonly'. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS2933)
1. ![MINOR][MINOR] [ContextValuesMarshaller.cs#L220](https://github.com/EvilBeaver/OneScript/blob/30d2725fe29dd66ce115b5e7d51649f0ebc30c95/src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs#L220): Replace this type-check-and-cast sequence with an 'as' and a null check. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS3247)
1. ![INFO][INFO] [XmlWriterImpl.cs#L321](https://github.com/EvilBeaver/OneScript/blob/30d2725fe29dd66ce115b5e7d51649f0ebc30c95/src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs#L321): Complete the task associated to this 'TODO' comment. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS1135)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by EvilBeaver** [comment]
Ну вот, после недавнего восстановления Сонар заработал

**Comment by nixel2007** [comment]
SonarQube analysis reported 5 issues

1. ![CRITICAL][CRITICAL] [ContextValuesMarshaller.cs#L53](https://github.com/EvilBeaver/OneScript/blob/38960cae84325abb36d86576b426173b1f9ef1f8/src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs#L53): Refactor this method to reduce its Cognitive Complexity from 20 to the 15 allowed. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS3776)
1. ![MAJOR][MAJOR] [ContextValuesMarshaller.cs#L217](https://github.com/EvilBeaver/OneScript/blob/38960cae84325abb36d86576b426173b1f9ef1f8/src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs#L217): Remove this useless assignment to local variable 'result'. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS1854)
1. ![MAJOR][MAJOR] [XmlWriterImpl.cs#L19](https://github.com/EvilBeaver/OneScript/blob/38960cae84325abb36d86576b426173b1f9ef1f8/src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs#L19): Fix this implementation of 'IDisposable' to conform to the dispose pattern. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS3881)
1. ![MAJOR][MAJOR] [XmlWriterImpl.cs#L25](https://github.com/EvilBeaver/OneScript/blob/38960cae84325abb36d86576b426173b1f9ef1f8/src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs#L25): Make '_nsmap' 'readonly'. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS2933)
1. ![MINOR][MINOR] [ContextValuesMarshaller.cs#L220](https://github.com/EvilBeaver/OneScript/blob/38960cae84325abb36d86576b426173b1f9ef1f8/src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs#L220): Replace this type-check-and-cast sequence with an 'as' and a null check. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS3247)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by nixel2007** [comment]
SonarQube analysis reported 5 issues

1. ![CRITICAL][CRITICAL] [ContextValuesMarshaller.cs#L53](https://github.com/EvilBeaver/OneScript/blob/2b30d49bf0c6ac02d8121ad31a078f01cc39152f/src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs#L53): Refactor this method to reduce its Cognitive Complexity from 20 to the 15 allowed. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS3776)
1. ![MAJOR][MAJOR] [ContextValuesMarshaller.cs#L217](https://github.com/EvilBeaver/OneScript/blob/2b30d49bf0c6ac02d8121ad31a078f01cc39152f/src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs#L217): Remove this useless assignment to local variable 'result'. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS1854)
1. ![MAJOR][MAJOR] [XmlWriterImpl.cs#L19](https://github.com/EvilBeaver/OneScript/blob/2b30d49bf0c6ac02d8121ad31a078f01cc39152f/src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs#L19): Fix this implementation of 'IDisposable' to conform to the dispose pattern. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS3881)
1. ![MAJOR][MAJOR] [XmlWriterImpl.cs#L25](https://github.com/EvilBeaver/OneScript/blob/2b30d49bf0c6ac02d8121ad31a078f01cc39152f/src/ScriptEngine.HostedScript/Library/Xml/XmlWriterImpl.cs#L25): Make '_nsmap' 'readonly'. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS2933)
1. ![MINOR][MINOR] [ContextValuesMarshaller.cs#L220](https://github.com/EvilBeaver/OneScript/blob/2b30d49bf0c6ac02d8121ad31a078f01cc39152f/src/ScriptEngine/Machine/Contexts/ContextValuesMarshaller.cs#L220): Replace this type-check-and-cast sequence with an 'as' and a null check. [rule](https://sonar.silverbulleters.org/coding_rules#rule_key=csharpsquid%3AS3247)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by dmpas** [comment]
@EvilBeaver сборку перезапустишь?

**Pull request merged**